internal/datastore/proxy/replicated.go

```diff
@@ -0,0 +1,341 @@
+package proxy
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	"github.com/authzed/spicedb/internal/datastore/common"
+	log "github.com/authzed/spicedb/internal/logging"
+	"github.com/authzed/spicedb/pkg/datastore"
+	"github.com/authzed/spicedb/pkg/datastore/options"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+// NewReplicatedDatastore creates a new datastore that writes to the provided primary and reads
+// from the provided replicas. The replicas are chosen in a round-robin fashion. If a replica does
+// not have the requested revision, the primary is used instead.
+//
+// NOTE: Be *very* careful when using this function. It is not safe to use this function without
+// knowledge of the layout of the underlying datastore and its replicas. See the notes below
+// on the requiresRevisionCheck parameter to understand the implications of using this function.
+//
+// If requiresRevisionCheck is true, the replica will be checked for the requested revision before
+// reading from it, which means that the read pool for the replicas *must* point to a *stable*
+// instance of the datastore (not a load balancer).
+//
+// If requiresRevisionCheck is false, the replica will be used for reads without checking for the
+// requested revision. This is useful when the read pool points to a load balancer that can
+// transparently handle the request. In this case, the primary will be used as a fallback if the
+// replica does not have the requested revision. The replica(s) supplied to this proxy *must*, therefore,
+// have strict read mode enabled, to ensure the query will fail with a RevisionUnavailableError if the
+// revision is not available.
+func NewReplicatedDatastore(primary datastore.Datastore, requiresRevisionCheck bool, replicas ...datastore.ReadOnlyDatastore) (datastore.Datastore, error) {
```

Comment:
It needs to be exposed by the CLI because it depends heavily on the layout of the replicas: if you're using something like pgbouncer and it can hit any replica, then strict read mode is a must. Since we can't determine that ourselves, it has to be a flag.

Reply:
Why would you want to run postgres in a mode other than strict?

Reply:
This is what I'm trying to bring up. This flags are not easy to understand and inconsistent between CLI and datastore proxies. The user will wonder when should I use one in favour of the other?

I assumed strict = the "joined" revision check, vs the isolated rev check before every other datastore query. I don't see a reason why postgres shouldn't be doing always the former, and hence I don't understand why the user should be making any call here, and question whether they have the information they need to make such call

Reply:
Because there will still be some *slight* overhead on the strict read mode approach and if you *know* your replica is one per pool, its safe to avoid using it. That being said, we could just remove the option and choose the "safe" one, so I think I'll do that